include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,159 @@
+        struct drop_last_bidirectional_view
+          : view_interface<drop_last_bidirectional_view<Rng>, range_cardinality<Rng>::value>
+        {
+            CONCEPT_ASSERT(ranges::BidirectionalRange<Rng>());
+        private:
+            friend range_access;
+            Rng rng;
+            std::size_t n;
+            detail::non_propagating_cache<sentinel_t<Rng>> sentinel;
+            void update_sentinel()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+            }
+        public:
+            drop_last_bidirectional_view() = default;
+            drop_last_bidirectional_view(Rng&& rng, std::size_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                update_sentinel();
+                return *sentinel;
+            }
+            Rng & base()
+            {
+                return rng;
+            }
+            Rng const & base() const
+            {
+                return rng;
+            }
+        };
+
+        template<typename Rng>
+        struct drop_last_forward_view
+          : view_adaptor<drop_last_forward_view<Rng>, Rng>
+        {
+            CONCEPT_ASSERT(ranges::ForwardRange<Rng>());
+        private:
+            friend range_access;
+            std::size_t n;
+
+            struct sentinel_adaptor;
+            class adaptor : public adaptor_base
+            {
+                friend sentinel_adaptor;
+                iterator_t<Rng> probe;
+            public:
+                adaptor() = default;
+                adaptor(const Rng& rng, std::size_t n)
+                {
+                    probe = ranges::next(ranges::begin(rng), n, ranges::end(rng));
+                }
+
+                void next(iterator_t<Rng>& it)
+                {
+                    ++it;
+                    ++probe;
+                }
+            };
+
+            struct sentinel_adaptor :  adaptor_base
+            {
+                bool empty(const iterator_t<Rng>&, const adaptor& ia, const sentinel_t<Rng>& s) const
+                {
+                    return ia.probe == s;
+                }
+            };
+
+            adaptor begin_adaptor() const { return {this->base(), n}; }
+            sentinel_adaptor end_adaptor() const { return {}; }
+        public:
+            drop_last_forward_view() = default;
+            drop_last_forward_view(Rng&& rng, std::size_t n)
+              : drop_last_forward_view::view_adaptor(std::move(rng))
+              , n(n)
+            {}
+        };
+
+        template<class Rng>
+        using drop_last_view =
+            meta::if_<BidirectionalRange<Rng>
+                , drop_last_bidirectional_view<Rng>
+                , drop_last_forward_view<Rng>>;
```

Comment:
Prefer to define a `drop_last_view` struct rather than an alias template. The alias template prevents us from optionally defining deduction guides.

Reply:
I didn't work with deduction guides. Does simple wrapper around that alias, will do the job?
```cpp
       template<class Rng>
       using drop_last_view_ =
           meta::if_<BidirectionalRange<Rng>
               , drop_last_bidirectional_view<Rng>
               , drop_last_forward_view<Rng>>;

        template<class Rng>
        struct drop_last_view : drop_last_view_<Rng>{
            using drop_last_view_<Rng>::drop_last_view_;
        };
```
Or I need to implement it as template specializations for `drop_last_view<class Rng, bool is_bidi>`?

_BTW, aren't this solved in C++20?_

Reply:
> I didn't work with deduction guides. Does simple wrapper around that alias, will do the job?

No. CTAD only works when the name of the template is a class template.
> _BTW, aren't this solved in C++20?_

Apparently, the CTAD wording for alias templates is complex. Hopefully it makes it in time for C++20. I wonder if they couldn't merge what they have and iterate on it until C++20 is published. Even then, there are the defect reports.